### ISSUE TITLE
Some code involving timestamps fail to respect columns'timestamp resolution

### ISSUE THREAD

**Issue by eyalroz** [rapidsai/cudf#395]
Some of our code compares or otherwise manipulates timestamps while ignoring the the fact that timestamps have a resolution, set at the column level.

Prominent (perhaps only?) example:

[`libgdf/src/gdf_table.cuh`, method `rows_equal`](https://github.com/rapidsai/cudf/blob/612ca60a7f41adb227ad556a91c76b69d2527d9b/libgdf/src/gdf_table.cuh#L581) - The GDF_TIMESTAMP case at [Line 676](https://github.com/rapidsai/cudf/blob/612ca60a7f41adb227ad556a91c76b69d2527d9b/libgdf/src/gdf_table.cuh#L676):
```
case GDF_TIMESTAMP:
{
    using col_type = int64_t;
    const col_type my_elem = static_cast<col_type*>(d_columns_data[i])[my_row_index];
    const col_type other_elem = static_cast<col_type*>(other.d_columns_data[i])[other_row_index];
    if(my_elem != other_elem)
        return false;
    break;
}
```
I've not yet determined whether any join-related code suffers from a similar problem, readers' assistance is requested. Thanks goes to @jrhemstad for noticing this.

**Comment by harrism** [comment]
What is "readers' assistance"? Do you have a suggestion for correct behavior?

**Comment by eyalroz** [comment]
> What is "readers' assistance"? 

People reading this bug. I'm wondering if this happens elsewhere.

> Do you have a suggestion for correct behavior? 

Either there's an enforcement of the same scale for timestamps (probably not the best thing to do), or the higher-resolution values get multiplied by the ratio of time unit lengths. And we would need to be careful about overflow.

> Can you provide a python example that gives incorrect results?

I do not have sufficient command of the Python part of cudf to do this.

**Comment by jrhemstad** [comment]
I'm of the opinion that timestamps with different resolutions should be treated as different types and are therefore not comparable for the purpose of something like `rows_equal`. 

To make this concrete, we could have `GDF_TIMESTAMP_S`, `GDF_TIMESTAMP_MS`, etc. in lieu of just `GDF_TIMESTAMP` with the gdf_dtype_info field that tells you the resolution.

**Comment by eyalroz** [comment]
... actually, it's worse than even the resolution. There's something we're not even supporting at all, yet, which is the timezone. See [here](https://arrow.apache.org/docs/cpp/classarrow_1_1_timestamp_type.html).

So, while @jrhemstad 's suggestion is consistent with the rest of our types, which have parameters 'encoded' in the enum, e.g. date32 vs date64 - I think this might not be a tenable approach in the long run. But this depends on how Arrow is used elsewhere (e.g. whether columns have a uniform non-UTC timezone set).

For now, the quickest fix would be to simply check for timestamp resolution between columns when comparing; and a second-quickest fix would be to add those types.

PS - I couldn't find timestamps mentioned in the Arrow [memory layout](https://arrow.apache.org/docs/memory_layout.html) documentation page, and they are barely mentioned on the [data types](https://arrow.apache.org/docs/python/data.html) documentation page.

**Comment by harrism** [comment]
WWPD? (What Would Pandas Do?)

**Comment by kkraus14** [comment]
Is this fixed in #654 @jrhemstad @eyalroz @harrism?

**Comment by jrhemstad** [comment]
> Is this fixed in #654 @jrhemstad @eyalroz @harrism?

Yes, I don't know why it didn't get automatically closed when it was merged. I have the "closes" keyword there.
